### ISSUE TITLE
A question about "using F-SAM under torch AMP"

### ISSUE THREAD

**Issue by QWTforGithub** [nblt/F-SAM#2]
Thanks for sharing the well organized code. May I ask how to use SAM under torch AMP (mixed precision training)?

**Comment by nblt** [comment]
Hi, thanks. I have not tried SAM under torch AMP. It seems that using closures is usually incompatible with mixed precision training. The following links may help:

https://discuss.pytorch.org/t/loss-scaling-and-custom-optimizers-cuda-mixed-precision/108400

https://docs.mosaicml.com/projects/composer/en/stable/method_cards/sam.html#how-to-use

**Comment by QWTforGithub** [comment]
> Hi, thanks. I have not tried SAM under torch AMP. It seems that using closures are usually incompatible with mixed precision training. The following links may help:
> 
> https://discuss.pytorch.org/t/loss-scaling-and-custom-optimizers-cuda-mixed-precision/108400
> 
> https://docs.mosaicml.com/projects/composer/en/stable/method_cards/sam.html#how-to-use

Thank you very much for your reply. I read the code of F-SAM carefully. I found that the network needs 2 forward propagation and 2 backward propagation at each iteration.  This usually increases the training time of the network, right?

**Comment by nblt** [comment]
Yes. Standard SAM requires two gradient backpropagations per iteration to solve the inner maximum problem. You can find more information in the original SAM [paper](https://arxiv.org/abs/2010.01412).

**Comment by QWTforGithub** [comment]
> Yes. Standard SAM requires two gradient backpropagations per iteration to solve the inner maximum problem. You can find more information in the original SAM [paper](https://arxiv.org/abs/2010.01412).

Thanks again for your reply. May I ask how the gradient visualization of SAM is drawn?

**Comment by nblt** [comment]
I guess you mean the visualization of the [loss landscape](https://github.com/marcellodebernardi/loss-landscapes).

**Comment by QWTforGithub** [comment]
> I guess you mean the visualization of the [loss landscape](https://github.com/marcellodebernardi/loss-landscapes).

Thank you very much!
